README.md
# Remove docker images

Remove obsolete docker images.

## Prerequisite

Install Docker.

## Supported tag syntax

The tags following this regex are supported:

    ^[vV]?(\d{1,5})(\.\d{1,5})?(\.\d{1,5})?(-.*)?$

ex) Supported tags:

    1.0
    1.0-SNAPSHOT
    v2.10.1
    2.0-BR291-SNAPSHOT
    8-jdk

ex) Non supported tag:

    123456.1 (major version has more than 5 digits)
    ver1.2

## Canonical version

If the tag matches the following regex, it is recognized as a canonical (non snapshot) version.

    ^[vV]?(\d{1,5})(\.\d{1,5})?(\.\d{1,5})?

## Snapshot version

If the tag ends with "-SNAPSHOT", it is recognized as a snapshot version.

## Branch

If non "-SNAPSHOT" string follows, it is recognized as a branch.

ex) Branch: BR102(canonical version)

    1.4-BR102

ex) Branch: BR102(snapshot version)

    1.4-BR102-SNAPSHOT

## How it works

1. Invoke "docker images" to obtain the list of docker image.

| Image name | tag |
-|-
| foo/myapp | 1.0 |
| foo/myapp | 1.0-SNAPSHOT |
| bar/app | 1.2 |
| foo/myapp | 1.1 |
| foo/myapp | 1.1-SNAPSHOT |
| foo/myapp | 1.2 |
| bar/app | 2.2 |
| foo/myapp | 1.3 |

2. Grouping the list by image name.

| Image name | tag |
-|-
| foo/myapp | 1.0 |
| foo/myapp | 1.0-SNAPSHOT |
| foo/myapp | 1.3 |
| foo/myapp | 1.1 |
| foo/myapp | 1.1-SNAPSHOT |
| foo/myapp | 1.2 |
| bar/app | 2.2 |
| bar/app | 1.2 |

3. Grouping the list by canonical/snapshot version.

| Image name | tag |
-|-
| foo/myapp | 1.0-SNAPSHOT |
| foo/myapp | 1.1-SNAPSHOT |
| foo/myapp | 1.0 |
| foo/myapp | 1.3 |
| foo/myapp | 1.1 |
| foo/myapp | 1.2 |
| bar/app | 2.2 |
| bar/app | 1.2 |

4. Sort by version number.

| Image name | tag |
-|-
| foo/myapp | 1.0-SNAPSHOT |
| foo/myapp | 1.1-SNAPSHOT |
| foo/myapp | 1.0 |
| foo/myapp | 1.1 |
| foo/myapp | 1.2 |
| foo/myapp | 1.3 |
| bar/app | 1.2 |
| bar/app | 2.2 |

5. Apply the following rule.<br/>
For snapshot version, keep the latest version. For canonical version, keep the three newest versions. You can modify the keep count by argument.

| Image name | tag | delete |
-|-|-
| foo/myapp | 1.0-SNAPSHOT | delete |
| foo/myapp | 1.1-SNAPSHOT | keep |
| foo/myapp | 1.0 | delete |
| foo/myapp | 1.1 | keep |
| foo/myapp | 1.2 | keep |
| foo/myapp | 1.3 | keep |
| bar/app | 1.2 | keep |
| bar/app | 2.2 | keep |

6. Invoke "docker rmi" to remove image

## Argument

- --version<br/>
Show tool version.

- --dry-run<br/>
Do not invoke "docker rmi" instead, just show the images that will be deleted.

- --keep [count]<br/>
Specify keep count for canonical versions. Default to 3.

- --keep-snapshot [count]<br/>
Specify keep count for snapshot versions. Default to 1.

- --repository-type [type]<br/>
Specify docker image repository type. Can use 'docker' and 'ibmcr'. The 'docker' means docker image repository. The 'ibmcr' means IBM Cloud Registry. Default to 'docker'.

## IBM Cloud Registry

If you want to use IBM Cloud Registry, specify '--repository-type ibmcr'. In this case, you need to login IBM Cloud Registry in advance.

    $ ibmcloud login
    $ ibmcloud cr login
    $ remove_docker_images --repository-type ibmcr

If you want to execute it in non-interactive environment such as in CI server, you can use API key.

    $ export IBMCLOUD_API_KEY=YOUR_CONTAINER_REGISTRY_APIKEY
    $ ibmcloud login
    $ ibmcloud cr login
    $ remove_docker_images --repository-type ibmcr

You can create API key at web console:

- Manage => IAM => Service IDs => Create
- Input name and click Create.
- Click Access Policies.
- Click Assign access.
- Select IAM services.
- Click drop down 'No access' and select Container Registry.
- In 'Service access', check Reader and Writer.
- Click Add.
- Click Assign.
- Click API keys.
- Click Create.
- Input Name and click Create.
- Now, you can copy the API Key.

## Binary

Linux(x86_64):
https://github.com/ruimo/remove_docker_images/releases


CHANGELOG.md
## [Unreleased] - ReleaseDate



src/arg.rs
extern crate clap;

use clap::{App, Arg};
use std::fmt;
use super::VERSION;
use super::docker_registry_type;

pub struct Args {
    pub is_dry_run: bool,
    pub keep_count: usize,
    pub keep_count_snapshot: usize,
    pub repository_type: docker_registry_type::DockerRegistryType,
    pub show_version: bool,
}

impl fmt::Display for Args {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "Args(is_dry_run: {}, keep_count: {}, keep_count_snapshot: {}, show_version: {})",
            self.is_dry_run, self.keep_count, self.keep_count_snapshot, self.show_version
        )
    }
}

fn parse_int(s: &str, var_name: &str) -> usize {
    match s.parse() {
        Result::Ok(val) => val,
        Result::Err(err) =>
            panic!("{} should be integer. {:?}", var_name, err),
    }
}
    
pub fn parse_arg() -> Args {
    let app = App::new("trimimages")
        .version(VERSION)
        .author("Shisei Hanai<ruimo.uno@gmail.com>")
        .about("Remove docker images")
        .arg(Arg::with_name("dryrun")
             .help("Dry run. Just show docker commands to execute.")
             .long("dry-run")
        )
        .arg(Arg::with_name("keep")
             .help("Keep count for canonical versioned image.")
             .long("keep")
             .default_value("3")
        )
        .arg(Arg::with_name("keep-snapshot")
             .help("Keep count for snapshot versioned image.")
             .long("keep-snapshot")
             .default_value("1")
        )
        .arg(Arg::with_name("repository-type")
             .help("Specify docker image repository type (docker or ibmcr).")
             .long("repository-type")
             .default_value("docker")
        )
        .arg(Arg::with_name("show-version")
             .help("Show version.")
             .long("version")
        )
        ;

    let matches = app.get_matches();

    Args {
        is_dry_run: matches.is_present("dryrun"),
        keep_count: parse_int(matches.value_of("keep").unwrap(), "keep count"),
        keep_count_snapshot: parse_int(matches.value_of("keep-snapshot").unwrap(), "keep count snapshot"),
        repository_type: docker_registry_type::parse(matches.value_of("repository-type").unwrap()).unwrap(),
        show_version: matches.is_present("show-version"),
    }
}


src/image.rs
use std::hash::{Hash, Hasher};

use std::collections::HashMap;
use std::collections::HashSet;
use std::fmt;

use super::version;

#[cfg(test)]
use super::image;

pub struct ImageEntry {
    pub id: String,
    pub ver: version::Version,
}

impl PartialEq for ImageEntry {
    fn eq(&self, other: &Self) -> bool {
        self.ver == other.ver
    }
}

impl Eq for ImageEntry {}

impl Hash for ImageEntry {
    fn hash<H:Hasher>(&self, state: &mut H) {
        self.ver.hash(state);
    }
}

impl fmt::Debug for ImageEntry {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "ImageEntry [id: {}, ver: {}]", self.id, self.ver)
    }
}

pub struct Images {
    // key: repository
    pub entries: HashMap<String, HashSet<ImageEntry>>
}

impl Images {
    pub fn delete<F>(&self, canonical_keep_count: usize, snapshot_keep_count: usize, mut del: F) -> ()
      where F : FnMut(&str, &version::Version) -> () // repo, version
    {
        for (repo, entry) in &self.entries {
            let mut sum_canonical: HashMap<&Option<String>, Vec<&version::Version>> = HashMap::new();
            let mut sum_snapshot: HashMap<&Option<String>, Vec<&version::Version>> = HashMap::new();

            for e in entry {
                let sum = if e.ver.is_snapshot { &mut sum_snapshot } else { &mut sum_canonical };
                let keep_count = if e.ver.is_snapshot { snapshot_keep_count } else {canonical_keep_count };
                let tbl = sum.entry(&e.ver.branch).or_insert_with(|| Vec::new());
                match tbl.binary_search(&&e.ver) {
                    Ok(_idx) => {
                    },
                    Err(idx) => {
                        tbl.insert(idx, &e.ver);
                    }
                }

                if keep_count < tbl.len() {
                    let v = tbl.remove(0);
                    del(repo, v);
                }
            }
        }
    }
}

#[test]
fn delete_test() {
    let parser = version::parser();
    let mut map: HashMap<String, HashSet<image::ImageEntry>> = HashMap::new();
    let mut entries0 = HashSet::new();
    entries0.insert(ImageEntry { id: "id00".to_string(), ver: parser.parse("1.0").unwrap() });
    entries0.insert(ImageEntry { id: "id01".to_string(), ver: parser.parse("1.1").unwrap() });
    entries0.insert(ImageEntry { id: "id02".to_string(), ver: parser.parse("1.10").unwrap() });
    entries0.insert(ImageEntry { id: "id03".to_string(), ver: parser.parse("1.2").unwrap() });

    entries0.insert(ImageEntry { id: "id04".to_string(), ver: parser.parse("1.2-SNAPSHOT").unwrap() });
    entries0.insert(ImageEntry { id: "id05".to_string(), ver: parser.parse("1.1-SNAPSHOT").unwrap() });

    entries0.insert(ImageEntry { id: "id06".to_string(), ver: parser.parse("1.2.0-BR123").unwrap() });
    entries0.insert(ImageEntry { id: "id07".to_string(), ver: parser.parse("1.2.1-BR123").unwrap() });
    entries0.insert(ImageEntry { id: "id08".to_string(), ver: parser.parse("1.2.10-BR123").unwrap() });
    entries0.insert(ImageEntry { id: "id09".to_string(), ver: parser.parse("1.2.2-BR123").unwrap() });

    entries0.insert(ImageEntry { id: "id10".to_string(), ver: parser.parse("1.2.2-BR123-SNAPSHOT").unwrap() });
    entries0.insert(ImageEntry { id: "id11".to_string(), ver: parser.parse("1.2.1-BR123-SNAPSHOT").unwrap() });

    map.insert("repo0".to_string(), entries0);

    let mut entries1 = HashSet::new();
    entries1.insert(ImageEntry { id: "id12".to_string(), ver: parser.parse("2.0").unwrap() });
    entries1.insert(ImageEntry { id: "id13".to_string(), ver: parser.parse("2.1").unwrap() });
    entries1.insert(ImageEntry { id: "id14".to_string(), ver: parser.parse("2.10").unwrap() });
    entries1.insert(ImageEntry { id: "id15".to_string(), ver: parser.parse("2.2").unwrap() });

    map.insert("repo1".to_string(), entries1);

    let images = Images { entries: map };

    let mut deleted = HashSet::new();
    images.delete(3, 1, |repo, ver| {
        deleted.insert(format!("{}:{}", repo, ver.to_string()));
    });

    assert_eq!(deleted.len(), 5);
    assert_eq!(deleted.contains("repo0:1.0"), true);
    assert_eq!(deleted.contains("repo0:1.1-SNAPSHOT"), true);
    assert_eq!(deleted.contains("repo0:1.2.0-BR123"), true);
    assert_eq!(deleted.contains("repo0:1.2.1-BR123-SNAPSHOT"), true);
    assert_eq!(deleted.contains("repo1:2.0"), true);
}


src/image_registry.rs
use std::process::{Command, Output};
use std::io::{BufRead, BufReader};
use std::collections::HashMap;
use std::collections::HashSet;

use super::version;
use super::image;

pub trait ImageRegistry {
    fn list(&self) -> image::Images;
    fn remove(&self, image_name: &str, ver: &version::Version, is_dryrun: bool) -> ();
}

pub struct DockerImageRegistry {
}

pub struct IbmCloudRegistry {
}

pub const DOCKER_IMAGE_REGISTRY: DockerImageRegistry = DockerImageRegistry {};

pub const IBM_CLOUD_REGISTRY: IbmCloudRegistry = IbmCloudRegistry {};

impl ImageRegistry for DockerImageRegistry {
    fn list(&self) -> image::Images {
        let out = Command::new("docker")
            .arg("images")
            .arg("--format")
            .arg("{{.ID}}\t{{.Repository}}\t{{.Tag}}")
            .output()
            .expect("Cannot run 'docker images'. Please check docker installation.");

        if out.status.success() {
            parse_docker_image_list(out)
        } else {
            println!("stdout: {}", String::from_utf8_lossy(&out.stdout));
            println!("stderr: {}", String::from_utf8_lossy(&out.stderr));
            panic!("docker images fails with status {}.", out.status);
        }
    }
    
    fn remove(&self, image_name: &str, ver: &version::Version, is_dryrun: bool) -> () {
        let img = format!("{}:{}", image_name, ver);

        if is_dryrun {
            println!("docker rmi {}", img);
        } else {
            let out = Command::new("docker")
                .arg("rmi")
                .arg(&img)
                .output()
                .expect("Cannot run 'docker rmi {}'. Please check docker installation.");

            if ! out.status.success() {
                println!("stdout: {}", String::from_utf8_lossy(&out.stdout));
                println!("stderr: {}", String::from_utf8_lossy(&out.stderr));
                panic!("docker rmi {} fails with status {}.", img, out.status);
            }
        }
    }
}

impl ImageRegistry for IbmCloudRegistry {
    fn list(&self) -> image::Images {
        let out = Command::new("ibmcloud")
            .arg("cr")
            .arg("images")
            .arg("--format")
            .arg("{{.Digest}}\t{{.Repository}}\t{{.Tag}}")
            .output()
            .expect("Cannot run 'ibmcloud cr images'. Please check ibmcloud CLI installation.");

        if out.status.success() {
            parse_docker_image_list(out)
        } else {
            println!("stdout: {}", String::from_utf8_lossy(&out.stdout));
            println!("stderr: {}", String::from_utf8_lossy(&out.stderr));
            panic!("ibmcloud cr images fails with status {}.", out.status);
        }
    }

    fn remove(&self, image_name: &str, ver: &version::Version, is_dryrun: bool) -> () {
        let img = format!("{}:{}", image_name, ver);

        if is_dryrun {
            println!("ibmcloud cr image-rm {}", img);
        } else {
            let out = Command::new("ibmcloud")
                .arg("cr")
                .arg("image-rm")
                .arg(&img)
                .output()
                .expect("Cannot run 'ibmcloud cr image-rm'. Please check ibmcloud CLI installation.");

            if ! out.status.success() {
                println!("stdout: {}", String::from_utf8_lossy(&out.stdout));
                println!("stderr: {}", String::from_utf8_lossy(&out.stderr));
                panic!("ibmcloud image-rm {} fails with status {}.", img, out.status);
            }
        }
    }
}

fn parse_docker_image_list(out: Output) -> image::Images {
    let br = BufReader::new(out.stdout.as_slice());
    let mut hash: HashMap<String, HashSet<image::ImageEntry>> = HashMap::new();
    let ver_parser = version::parser();
        
    for (_, line) in br.lines().enumerate() {
        let l = line.unwrap();
        let mut z = l.split('\t');
        let id = z.next().unwrap();
        let repository = z.next().unwrap();
        let tag = z.next().unwrap();
        match ver_parser.parse(tag) {
            None => println!("Version(={}) is unrecognized ignored: {}", tag, l),
            Some(ver) => {
                match hash.get_mut(repository) {
                    Some(entry) => {
                        entry.insert(image::ImageEntry {id: id.to_string(), ver: ver});
                    },
                    None => {
                        let mut set = HashSet::new();
                        set.insert(image::ImageEntry {id: id.to_string(), ver: ver});
                        hash.insert(repository.to_string(), set);
                    }
                }
            }
        }
    }
        
    image::Images {entries: hash}
}


src/version.rs
extern crate regex;

use std::cmp::Ordering;
use std::hash::{Hash, Hasher};
use std::fmt;

pub struct Version {
    pub major: i32,
    pub minor: i32,  // If minor version does not exist, minor = -1
    pub patch: i32,  // If patch version does not exist, patch = -1
    pub is_snapshot: bool,
    pub branch: Option<String>,
    pub raw: String,
}

pub struct VersionParser {
    re: regex::Regex
}

pub fn parser() -> VersionParser {
    VersionParser {
        re: regex::Regex::new(r"^[vV]?(\d{1,5})(\.\d{1,5})?(\.\d{1,5})?(-.*)?$").unwrap()
    }
}

impl VersionParser {
    pub fn parse(&self, s: &str) -> Option<Version> {
        self.re.captures(s).map(|caps| {
            let minor = match caps.get(2) {
                Some(m) => m.as_str()[1..].parse().unwrap(),
                None => -1
            };

            let patch = match caps.get(3) {
                Some(m) => m.as_str()[1..].parse().unwrap(),
                None => -1
            };

            let (branch, is_snapshot) = match caps.get(4) {
                Some(m) => {
                    let v: Vec<&str> = m.as_str().split("-SNAPSHOT").collect();
                    let br = v[0].to_string();
                    (if br.len() > 1 {Some(br[1..].to_string())} else {None}, v.len() == 2)
                },
                None => (None, false)
            };


            Version {
                major: caps.get(1).unwrap().as_str().parse().unwrap(),
                minor: minor,
                patch: patch,
                is_snapshot: is_snapshot,
                branch: branch,
                raw: s.to_string(),
            }
        })
    }
}

impl PartialEq for Version {
    fn eq(&self, other: &Self) -> bool {
        self.major == other.major &&
            self.minor == other.minor &&
            self.patch == other.patch &&
            self.is_snapshot == other.is_snapshot &&
            self.branch == other.branch
    }
}

impl Eq for Version {}

impl Hash for Version {
    fn hash<H:Hasher>(&self, state: &mut H) {
        self.major.hash(state);
        self.minor.hash(state);
        self.patch.hash(state);
        self.is_snapshot.hash(state);
        self.branch.hash(state);
    }
}

impl fmt::Display for Version {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         write!(f, "{}", self.raw)
    }
}

impl fmt::Debug for Version {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Version [major: {}, minor: {}, patch: {}, branch: {:?}, is_snapshot: {}]",
               self.major, self.minor, self.patch, self.branch, self.is_snapshot)
    }
}

impl Ord for Version {
    fn cmp(&self, other: &Self) -> Ordering {
        let is = self.is_snapshot.cmp(&other.is_snapshot);
        if is != Ordering:: Equal {
            return is;
        }

        let br = self.branch.cmp(&other.branch);
        if br != Ordering::Equal {
            return br;
        }

        let ma = self.major.cmp(&other.major);
        if ma != Ordering::Equal {
            return ma;
        }

        let mi = self.minor.cmp(&other.minor);
        if mi != Ordering::Equal {
            return mi;
        }

        self.patch.cmp(&other.patch)
    }
}

impl PartialOrd for Version {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

#[test]
fn test_major_minor_patch() {
    let v = parser().parse("1.2.3").unwrap();
    assert_eq!(v.major, 1);
    assert_eq!(v.minor, 2);
    assert_eq!(v.patch, 3);
    assert_eq!(v.is_snapshot, false);
    assert_eq!(v.branch, None);
}

#[test]
fn test_major_minor_patch_snapshot() {
    let v = parser().parse("1.2.3-SNAPSHOT").unwrap();
    assert_eq!(v.major, 1);
    assert_eq!(v.minor, 2);
    assert_eq!(v.patch, 3);
    assert_eq!(v.is_snapshot, true);
    assert_eq!(v.branch, None);
}

#[test]
fn test_major_minor() {
    let v = parser().parse("1.2").unwrap();
    assert_eq!(v.major, 1);
    assert_eq!(v.minor, 2);
    assert_eq!(v.patch, -1);
    assert_eq!(v.is_snapshot, false);
    assert_eq!(v.branch, None);
}

#[test]
fn test_major() {
    let v = parser().parse("1").unwrap();
    assert_eq!(v.major, 1);
    assert_eq!(v.minor, -1);
    assert_eq!(v.patch, -1);
    assert_eq!(v.is_snapshot, false);
    assert_eq!(v.branch, None);
}

#[test]
fn test_branch() {
    let v = parser().parse("1.2-BR102").unwrap();
    assert_eq!(v.major, 1);
    assert_eq!(v.minor, 2);
    assert_eq!(v.patch, -1);
    assert_eq!(v.is_snapshot, false);
    assert_eq!(v.branch, Some("BR102".to_string()));
}

#[test]
fn test_branch_snapshot() {
    let v = parser().parse("1.2-BR102-SNAPSHOT").unwrap();
    assert_eq!(v.major, 1);
    assert_eq!(v.minor, 2);
    assert_eq!(v.patch, -1);
    assert_eq!(v.is_snapshot, true);
    assert_eq!(v.branch, Some("BR102".to_string()));
}

#[test]
fn latest() {
    assert_eq!(parser().parse("latest").is_none(), true);
}

#[test]
fn v() {
    let v = parser().parse("v0.31.2").unwrap();
    assert_eq!(v.major, 0);
    assert_eq!(v.minor, 31);
    assert_eq!(v.patch, 2);
    assert_eq!(v.is_snapshot, false);
    assert_eq!(v.branch, None);
    assert_eq!(format!("{}", v), "v0.31.2");
}

#[test]
fn sdk8() {
    let v = parser().parse("8-sdk").unwrap();
    assert_eq!(v.major, 8);
    assert_eq!(v.minor, -1);
    assert_eq!(v.patch, -1);
    assert_eq!(v.is_snapshot, false);
    assert_eq!(v.branch, Some("sdk".to_string()));
}


#[test]
fn major_minor_patch_cmp() {
    let parser = parser();
    assert_eq!(parser.parse("1.2.3").unwrap().cmp(&parser.parse("1.2.3").unwrap()), Ordering::Equal);
    assert_eq!(parser.parse("1.2.2").unwrap().cmp(&parser.parse("1.2.3").unwrap()), Ordering::Less);
    assert_eq!(parser.parse("1.2.3").unwrap().cmp(&parser.parse("1.2.2").unwrap()), Ordering::Greater);
    assert_eq!(parser.parse("1.2.10").unwrap().cmp(&parser.parse("1.2.3").unwrap()), Ordering::Greater);

    assert_eq!(parser.parse("1.2").unwrap().cmp(&parser.parse("1.2").unwrap()), Ordering::Equal);
    assert_eq!(parser.parse("1.2").unwrap().cmp(&parser.parse("1.1").unwrap()), Ordering::Greater);
    assert_eq!(parser.parse("1.2").unwrap().cmp(&parser.parse("1.3").unwrap()), Ordering::Less);

    assert_eq!(parser.parse("1").unwrap().cmp(&parser.parse("1").unwrap()), Ordering::Equal);
    assert_eq!(parser.parse("1").unwrap().cmp(&parser.parse("2").unwrap()), Ordering::Less);
    assert_eq!(parser.parse("2").unwrap().cmp(&parser.parse("1").unwrap()), Ordering::Greater);
}

#[test]
fn major_minor_patch_cmp_snapshot() {
    let parser = parser();
    assert_eq!(parser.parse("1.2.3-SNAPSHOT").unwrap().cmp(&parser.parse("1.2.3-SNAPSHOT").unwrap()), Ordering::Equal);
    assert_eq!(parser.parse("1.2.2-SNAPSHOT").unwrap().cmp(&parser.parse("1.2.3-SNAPSHOT").unwrap()), Ordering::Less);
    assert_eq!(parser.parse("1.2.3-SNAPSHOT").unwrap().cmp(&parser.parse("1.2.2-SNAPSHOT").unwrap()), Ordering::Greater);
    assert_eq!(parser.parse("1.2.10-SNAPSHOT").unwrap().cmp(&parser.parse("1.2.2-SNAPSHOT").unwrap()), Ordering::Greater);

    assert_eq!(parser.parse("1.2-SNAPSHOT").unwrap().cmp(&parser.parse("1.2-SNAPSHOT").unwrap()), Ordering::Equal);
    assert_eq!(parser.parse("1.2-SNAPSHOT").unwrap().cmp(&parser.parse("1.1-SNAPSHOT").unwrap()), Ordering::Greater);
    assert_eq!(parser.parse("1.2-SNAPSHOT").unwrap().cmp(&parser.parse("1.3-SNAPSHOT").unwrap()), Ordering::Less);

    assert_eq!(parser.parse("1-SNAPSHOT").unwrap().cmp(&parser.parse("1-SNAPSHOT").unwrap()), Ordering::Equal);
    assert_eq!(parser.parse("1-SNAPSHOT").unwrap().cmp(&parser.parse("2-SNAPSHOT").unwrap()), Ordering::Less);
    assert_eq!(parser.parse("2-SNAPSHOT").unwrap().cmp(&parser.parse("1-SNAPSHOT").unwrap()), Ordering::Greater);

    assert_eq!(parser.parse("2").unwrap().cmp(&parser.parse("2-SNAPSHOT").unwrap()) != Ordering::Equal, true);
}

#[test]
fn branch() {
    let parser = parser();
    assert_eq!(parser.parse("1.2.3-BR123").unwrap().cmp(&parser.parse("1.2.3-BR123").unwrap()), Ordering::Equal);
    assert_eq!(parser.parse("1.2.2-BR123").unwrap().cmp(&parser.parse("1.2.3-BR123").unwrap()), Ordering::Less);
    assert_eq!(parser.parse("1.2.3-BR123").unwrap().cmp(&parser.parse("1.2.2-BR123").unwrap()), Ordering::Greater);
    assert_eq!(parser.parse("1.2.10-BR123").unwrap().cmp(&parser.parse("1.2.2-BR123").unwrap()), Ordering::Greater);

    assert_eq!(parser.parse("1.2-BR123").unwrap().cmp(&parser.parse("1.2-BR123").unwrap()), Ordering::Equal);
    assert_eq!(parser.parse("1.2-BR123").unwrap().cmp(&parser.parse("1.1-BR123").unwrap()), Ordering::Greater);
    assert_eq!(parser.parse("1.2-BR123").unwrap().cmp(&parser.parse("1.3-BR123").unwrap()), Ordering::Less);

    assert_eq!(parser.parse("1-BR123").unwrap().cmp(&parser.parse("1-BR123").unwrap()), Ordering::Equal);
    assert_eq!(parser.parse("1-BR123").unwrap().cmp(&parser.parse("2-BR123").unwrap()), Ordering::Less);
    assert_eq!(parser.parse("2-BR123").unwrap().cmp(&parser.parse("1-BR123").unwrap()), Ordering::Greater);

    assert_eq!(parser.parse("2").unwrap().cmp(&parser.parse("2-BR123").unwrap()) != Ordering::Equal, true);
}

#[test]
fn branch_snapshot() {
    let parser = parser();
    assert_eq!(parser.parse("1.2.3-BR123-SNAPSHOT").unwrap().cmp(&parser.parse("1.2.3-BR123-SNAPSHOT").unwrap()), Ordering::Equal);
    assert_eq!(parser.parse("1.2.2-BR123-SNAPSHOT").unwrap().cmp(&parser.parse("1.2.3-BR123-SNAPSHOT").unwrap()), Ordering::Less);
    assert_eq!(parser.parse("1.2.3-BR123-SNAPSHOT").unwrap().cmp(&parser.parse("1.2.2-BR123-SNAPSHOT").unwrap()), Ordering::Greater);
    assert_eq!(parser.parse("1.2.10-BR123-SNAPSHOT").unwrap().cmp(&parser.parse("1.2.2-BR123-SNAPSHOT").unwrap()), Ordering::Greater);

    assert_eq!(parser.parse("1.2-BR123-SNAPSHOT").unwrap().cmp(&parser.parse("1.2-BR123-SNAPSHOT").unwrap()), Ordering::Equal);
    assert_eq!(parser.parse("1.2-BR123-SNAPSHOT").unwrap().cmp(&parser.parse("1.1-BR123-SNAPSHOT").unwrap()), Ordering::Greater);
    assert_eq!(parser.parse("1.2-BR123-SNAPSHOT").unwrap().cmp(&parser.parse("1.3-BR123-SNAPSHOT").unwrap()), Ordering::Less);

    assert_eq!(parser.parse("1-BR123-SNAPSHOT").unwrap().cmp(&parser.parse("1-BR123-SNAPSHOT").unwrap()), Ordering::Equal);
    assert_eq!(parser.parse("1-BR123-SNAPSHOT").unwrap().cmp(&parser.parse("2-BR123-SNAPSHOT").unwrap()), Ordering::Less);
    assert_eq!(parser.parse("2-BR123-SNAPSHOT").unwrap().cmp(&parser.parse("1-BR123-SNAPSHOT").unwrap()), Ordering::Greater);

    assert_eq!(parser.parse("2-BR123").unwrap().cmp(&parser.parse("2-BR123-SNAPSHOT").unwrap()) != Ordering::Equal, true);
}
